### PR TITLE
Use simplify_exprtt::resultt in pre-order simplification steps

### PR DESCRIPTION
**Pull request by tautschnig**
The use of resultt increases type safety as the expression to be
simplified is no longer modified in place. All post-order simplification
steps already use resultt, but pre-order steps had been left to be
done.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]

File: src/util/simplify_expr_if.cpp [hunk 1]
```diff
@@ -260,21 +279,18 @@ bool simplify_exprt::simplify_if_preorder(if_exprt &expr)
...
 
-    auto r_truevalue = simplify_rec(truevalue);
-    if(r_truevalue.has_changed())
-    {
-      truevalue = r_truevalue.expr;
-      no_change = false;
-    }
+    auto r_truevalue = simplify_rec(swap_branches ? falsevalue : truevalue);
+    if(swap_branches)
+      r_truevalue.expr_changed = CHANGED;
```
Comment (kroening) [inline src/util/simplify_expr_if.cpp:285]: Why is this needed? The if-expression will be marked as changed anyway.

File: src/util/simplify_expr_if.cpp [hunk 2]
```diff
@@ -283,48 +299,36 @@ bool simplify_exprt::simplify_if_preorder(if_exprt &expr)
...
 
-    auto r_falsevalue = simplify_rec(falsevalue);
-    if(r_falsevalue.has_changed())
-    {
-      falsevalue = r_falsevalue.expr;
-      no_change = false;
-    }
+    auto falsevalue = simplify_rec(swap_branches ? falsevalue : truevalue);
+    if(swap_branches)
+      r_falsevalue.expr_changed = CHANGED;
```
Comment (kroening) [inline src/util/simplify_expr_if.cpp:303]: Same here

File: src/util/simplify_expr_if.cpp (hunk 1 above)
Comment (tautschnig) [reply src/util/simplify_expr_if.cpp:285]: Yes, but `build_if_expr` would not actually replace the `truevalue`.

File: src/util/simplify_expr_if.cpp (hunk 2 above)
Comment (tautschnig) [reply src/util/simplify_expr_if.cpp:303]: `build_if_expr` would not actually replace the `falsevalue`.

File: src/util/simplify_expr_if.cpp (hunk 1 above)
Comment (tautschnig) [reply src/util/simplify_expr_if.cpp:285]: I have refactored the code and added a comment.

**Review by kroening — APPROVED** [review]
